include/ttmlir/Dialect/TTNN/Transforms/Workarounds/Decomposition/ReduceOpsRewritePattern.h

```diff
@@ -123,6 +125,62 @@ class ReduceOpsKeepDimRewritePattern : public OpRewritePattern<ReduceOp> {
     return false;
   }
 };
+
+template <typename ReduceOp>
+class ReduceOpsPadInputRewritePattern : public OpRewritePattern<ReduceOp> {
+public:
+  using OpRewritePattern<ReduceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ReduceOp srcOp,
+                                PatternRewriter &rewriter) const override {
+    auto inputType = mlir::cast<RankedTensorType>(srcOp.getInput().getType());
+    llvm::SmallVector<int64_t> reductionDims = getReduceDims(srcOp.getDimArg());
+    llvm::ArrayRef<int64_t> shape = inputType.getShape();
+    auto newShape = llvm::SmallVector<int64_t>(shape.begin(), shape.end());
+    llvm::SmallVector<int32_t> paddingArray(2 * shape.size(), 0);
+
+    constexpr int tileSize = 32;
+    for (auto dim : reductionDims) {
+      if (dim < 0) {
+        dim += shape.size();
+      }
+      auto padded_size = (shape[dim] + tileSize - 1) / tileSize * tileSize;
+      newShape[dim] = padded_size;
+      paddingArray[2 * dim + 1] = padded_size - shape[dim];
+    }
+    if (llvm::all_of(paddingArray, [](int32_t i) { return i == 0; })) {
+      return failure();
+    }
+
+    auto resultType = RankedTensorType::get(
+        newShape, inputType.getElementType(), inputType.getEncoding());
+    float paddingValue = getPaddingValue();
+    auto padOp = rewriter.create<mlir::tt::ttnn::PadOp>(
+        srcOp.getLoc(), resultType, srcOp.getInput(),
+        rewriter.getDenseI32ArrayAttr(paddingArray),
+        rewriter.getFloatAttr(rewriter.getF32Type(), APFloat(paddingValue)),
+        /* use_multicore */ rewriter.getBoolAttr(true),
+        /* memory_config*/ nullptr);
+
+    rewriter.replaceOpWithNewOp<ReduceOp>(
+        srcOp, srcOp.getResult().getType(), padOp.getResult(),
+        rewriter.getBoolAttr(srcOp.getKeepDim()),
+        srcOp.getDimArg().value_or(nullptr));
```

Comment:
```suggestion
    rewriter.modifyOpInPlace(srcOp, [&]() {
      srcOp.getInputMutable().assign(padOp);
    });
```
Might be preferable in this case.

Reply:
TIL that exists. I agree it's preferable.